platforms/posix/src/lockstep_scheduler/src/lockstep_scheduler.cpp

```diff
@@ -1,102 +1,123 @@
 #include "lockstep_scheduler/lockstep_scheduler.h"
 
-
-uint64_t LockstepScheduler::get_absolute_time() const
+LockstepScheduler::~LockstepScheduler()
 {
-	return time_us_;
+	// cleanup the linked list
+	std::unique_lock<std::mutex> lock_timed_waits(_timed_waits_mutex);
+
+	while (_timed_waits) {
+		TimedWait *tmp = _timed_waits;
+		_timed_waits = _timed_waits->next;
+		tmp->removed = true;
+	}
 }
 
 void LockstepScheduler::set_absolute_time(uint64_t time_us)
 {
-	time_us_ = time_us;
+	_time_us = time_us;
 
 	{
-		std::unique_lock<std::mutex> lock_timed_waits(timed_waits_mutex_);
+		std::unique_lock<std::mutex> lock_timed_waits(_timed_waits_mutex);
+		_setting_time = true;
 
-		auto it = std::begin(timed_waits_);
+		TimedWait *timed_wait = _timed_waits;
+		TimedWait *timed_wait_prev = nullptr;
 
-		while (it != std::end(timed_waits_)) {
+		while (timed_wait) {
+			// Clean up the ones that are already done from last iteration.
+			if (timed_wait->done) {
+				// Erase from the linked list
+				if (timed_wait_prev) {
+					timed_wait_prev->next = timed_wait->next;
 
-			std::shared_ptr<TimedWait> temp_timed_wait = *it;
+				} else {
+					_timed_waits = timed_wait->next;
+				}
 
-			// Clean up the ones that are already done from last iteration.
-			if (temp_timed_wait->done) {
-				it = timed_waits_.erase(it);
+				TimedWait *tmp = timed_wait;
+				timed_wait = timed_wait->next;
+				tmp->removed = true;
 				continue;
 			}
 
-			if (temp_timed_wait->time_us <= time_us &&
-			    !temp_timed_wait->timeout &&
-			    !temp_timed_wait->done) {
-				temp_timed_wait->timeout = true;
+			if (timed_wait->time_us <= time_us &&
+			    !timed_wait->timeout) {
```

Comment:
Are you sure that we don't need `&& !temp_timed_wait->done` anymore?

Reply:
No, since we check for that above already. If it were needed here, it would mean we have a race condition.